Sources/GRPCHTTP2Core/Client/Resolver/NameResolver.swift

```diff
@@ -0,0 +1,117 @@
+    /// Addresses are pushed to the resolve by an external source.
+    case push
+    /// Addresses are resolved lazily, when the caller asks them to be resolved.
+    case pull
+  }
+
+  /// Create a new name resolver.
+  public init(results: RPCAsyncSequence<NameResolutionResult>, updateMode: UpdateMode) {
+    self.names = results
+    self.updateMode = updateMode
+  }
+}
+
+/// The result of name resolution, a service of endpoints to connection to and the service
+/// configuration reported by the resolver.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public struct NameResolutionResult: Hashable, Sendable {
+  /// A list of endpoints to connect to.
+  public var endpoints: [Endpoint]
+
+  /// The service configuration reported by the resolver, or an error if it couldn't be parsed.
+  public var serviceConfiguration: Result<ServiceConfiguration, RPCError>
+
+  public init(
+    endpoints: [Endpoint],
+    serviceConfiguration: Result<ServiceConfiguration, RPCError>
+  ) {
+    self.endpoints = endpoints
+    self.serviceConfiguration = serviceConfiguration
+  }
+}
+
+/// A group of addresses which are considered equivalent when establishing a connection.
+public struct Endpoint: Hashable, Sendable {
+  /// A list of equivalent addresses.
+  ///
+  /// Earlier addresses are typically but not always connected to first. Some load balancers may
+  /// choose to ignore the order.
+  public var addresses: [SocketAddress]
+
+  /// Create a new ``Endpoint``.
+  /// - Parameter addresses: A list of equivalent addresses.
+  public init(addresses: [SocketAddress]) {
+    self.addresses = addresses
+  }
+}
+
+/// A resolver capable of resolving targets of type ``Target``.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public protocol NameResolverFactory<Target> {
+  /// The type of ``ResolvableTarget`` this factory makes resolvers for.
+  associatedtype Target: ResolvableTarget
+
+  /// Creates a resolver for the given target.
+  ///
+  /// - Parameter target: The target to make a resolver for.
+  /// - Returns: The name resolver for the target.
+  func resolver(for target: Target) -> NameResolver
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+extension NameResolverFactory {
+  /// Returns whether the given target is compatible with this factory.
+  ///
+  /// - Parameter target: The target to check the compatibility of.
+  /// - Returns: Whether the target is compatible with this factory.
+  func isCompatible<Other: ResolvableTarget>(withTarget target: Other) -> Bool {
+    return target is Target
+  }
+
+  /// Returns a name resolver if the given target is compatible.
+  ///
+  /// - Parameter target: The target to make a name resolver for.
+  /// - Returns: A name resolver or `nil` if the target isn't compatible.
+  func makeResolverIfCompatible<Other: ResolvableTarget>(_ target: Other) -> NameResolver? {
+    guard let target = target as? Target else { return nil }
```

Comment:
Maybe worth using the above func?
```suggestion
    guard self.isCompatible(withTarget: target) else { return nil }
```

Reply:
If we did use `isCompatible(withTarget:)` we'd need to call `self.resolver(for: target as! Target)` and I'd rather avoid the `!` here.